README.md
# invrementalParallelCanTree

history.txt
spark-submit', '--class', 'CanTreeMain', '--master', 'spark://cn41-ib:7077', '--driver-memory', '8│379102 kulev     20   0  171012   2396   1204 R   0.7  0.0 128:47.84 top                         
g', '--executor-memory', '80g', '--num-executors', '7', '--executor-cores', '20', '--files', 'src/m│272424 kulev     20   0  200444   2376   1012 S   0.0  0.0   1:11.08 sshd                        
ain/resources/log4j_file_executor.properties', '--conf', 'spark.driver.extraJavaOptions=-Dlog4j.conf│272425 kulev     20   0  127720    676    672 S   0.0  0.0   0:00.12 bash                        
iguration=file:src/main/resources/log4j_file.properties -Xss1g', '--conf', 'spark.executor.extraJava│379481 kulev     20   0   15.2g 572260   7632 S   0.0  0.3  17:14.71 java                        
Options=-Dlog4j.configuration=file:src/main/resources/log4j_file_executor.properties -Xss1g', 'targe│                                                                                                 
t/scala-2.11/cantree_2.11-0.1.jar', '--num-partitions', '1000', '--min-support', '0.01', '--in-file-│                                                                                                 
list-path', '/RG/rg-gudes/kulev/dataset/T15I5D1000N10_20splits/filesList1.txt', '--app-name', 'MINMI│                                                                                                 
N_FREQ_T15D1M10K_1000_20_7_80g_0_01', '--freq-sort', '1', '--min-min-support', '0.1



spark-submit --class CanTreeMain --master spark://cn41-ib:7077 --driver-memory 80g --executor-memory 80g --num-executors 7 --executor-cores 20 --files src/main/resources/log4j_file_executor.properties --conf 'spark.driver.extraJavaOptions=-Dlog4j.configuration=file:src/main/resources/log4j_file.properties -Xss1g' --conf 'spark.executor.extraJavaOptions=-Dlog4j.configuration=file:src/main/resources/log4j_file_executor.properties -Xss1g' target/scala-2.11/cantree_2.11-0.1.jar --num-partitions 1000 --min-support 0.01 --in-file-list-path /RG/rg-gudes/kulev/dataset/T15I5D1000N10_20splits/filesList1.txt  --app-name MINMIN_FREQ_T15D1M10K_1000_20_7_80g_0_01 --freq-sort 1 --min-min-support 0.1






def checkContains(arr : Array[Array[Int]], item : Int) : Boolean = {
	for (i <- arr) 
		for (j <- i)
			if (j==item)
				return true
	return false
}

def findMissing(arr1 : Array[org.apache.spark.mllib.fpm.FPGrowth.FreqItemset[Int]], arr2 : Array[levko.cantree.utils.CanTreeFPGrowth.FreqItemset[Int]]) : org.apache.spark.mllib.fpm.FPGrowth.FreqItemset[Int] = {
	var res = null;
	var counter = 0;
	for (item1 <- arr1) {
		var found = false
		if (item1.items.size <=1) {
			for (item2 <- arr2) {
				if (item2.items.size <=1) {
					if (item1.items(0)==item2.items(0)) {
						//println("found "+item1.items(0))
						found = true
						counter+=1
						println(counter +" found "+item1.items(0))
					}
				}
			}
			if (!found) {
				return item1
			}
		}
	}
	return res
}




data.flatMap { transaction =>
    genCondTransactions(transaction, itemToRank, partitioner)
  }.aggregateByKey(new CanTreeV1[Int], partitioner.numPartitions)(
    (tree, transaction) => tree.add(transaction, 1L),
    (tree1, tree2) => tree1.merge(tree2))




import levko.cantree.utils.{CanTreeFPGrowth, CanTreeV1}
import org.apache.spark.{HashPartitioner, Partitioner, SparkException}
import org.apache.spark.sql.{DataFrame, SparkSession}
import org.apache.spark.rdd.RDD
import org.apache.spark.rdd.RDD._
import scala.collection.mutable
import scala.collection.mutable.ListBuffer
import org.apache.spark.sql.functions.{col, collect_list, min}
import scala.reflect.ClassTag
import org.apache.spark.sql.types.{IntegerType, StructField, StructType}
import org.apache.log4j.Logger
import org.apache.spark.mllib.fpm.FPGrowth
import scala.io.Source
import scala.sys.exit
  type Sorter[T] = (T, T) => Boolean
  type StringSorter = Sorter[String]
  type IntSorter = Sorter[Int]
  def stringSorter : StringSorter = (s1,s2) => s1<s2
  def intSorter : IntSorter = (i1,i2) => i1<i2
  def customSorter(map: Map[Int, Long]): Sorter[Int] = { (i1, i2) =>
    (map.get(i1), map.get(i2)) match {
      case (Some(ic1), Some(ic2)) => ic1 < ic2
      case(Some(ic1),_)           => false
      case(_,Some(ic2))           => true
      case _                      => i1 < i2
    }
  }
  def prepareTransactions[Item](filePath: String, spark: SparkSession, customSchema : StructType):  DataFrame = {
    val df = spark.read.format("csv").option("header", "false").schema(customSchema).load(filePath)
    df.groupBy("InvoiceNo").agg(collect_list(col("StockCode")))
  }
  def getItemsCount[Item: ClassTag](data: RDD[Array[Item]]): RDD[(Item, Long)] = {
    data.flatMap { t =>
      val uniq = t.toSet
      if (t.length != uniq.size) {
        throw new SparkException(s"Items in a transaction must be unique but got ${t.toSeq}.")
      }
      t
    }.map(v => (v, 1L))
      .reduceByKey(_ + _)
  }
  def prepareTransactions[Item](filePath: String, spark: SparkSession, customSchema : StructType):  DataFrame = {
    val df = spark.read.format("csv").option("header", "true").schema(customSchema).load(filePath)
    df.groupBy("InvoiceNo").agg(collect_list(col("StockCode")))
  }
val numPartitions = 3
val minSupport = 0.1
val minminSupport = 0.8
    val customSchema = StructType(Array(
      StructField("InvoiceNo", IntegerType, true),
      StructField("StockCode", IntegerType, true))
    )
    val partitioner : HashPartitioner = new HashPartitioner(numPartitions)
val f = "/home/lev/Documents/teza/cantree/smallTest1.csv"
val dfGrouped = prepareTransactions(f,spark,customSchema)
      val transactions = dfGrouped.rdd.map(t=>t(1).asInstanceOf[mutable.WrappedArray[Int]].toArray)
      val minMinSupport = math.floor(transactions.count()*minSupport*minminSupport).toLong
      val model = new CanTreeFPGrowth().setMinSupport(minSupport).setPartitioner(partitioner).setMinMinSupport(minMinSupport)
        val countMap = transactions.flatMap { t =>

runTestcases.py
import os, shlex, subprocess
import argparse

def prepareLogFile(logFilePath,errorFilePath,outputFilePath):
    log4jProperties = 'log4j.rootCategory=INFO,FILE,R' + os.linesep + \
                      'log4j.appender.FILE=org.apache.log4j.FileAppender'+ os.linesep + \
                      'log4j.appender.FILE.File={logFilePath}' + os.linesep + \
                      'log4j.appender.FILE.ImmediateFlush=true' + os.linesep + \
                      'log4j.appender.FILE.MaxFileSize=10MB' + os.linesep + \
                      'log4j.appender.FILE.MaxBackupIndex=10' + os.linesep + \
                      'log4j.appender.FILE.layout=org.apache.log4j.PatternLayout' + os.linesep + \
                      'log4j.appender.FILE.layout.conversionPattern=%m%n' + os.linesep + \
                      'log4j.appender.R=org.apache.log4j.RollingFileAppender' + os.linesep + \
                      'log4j.appender.R.File={errorFilePath}' + os.linesep + \
                      'log4j.appender.R.MaxFileSize=10MB' + os.linesep + \
                      'log4j.appender.R.layout=org.apache.log4j.PatternLayout' + os.linesep + \
                      'log4j.appender.R.layout.ConversionPattern=%p %t %c - %m%n' + os.linesep + \
                      'log4j.logger.org.spark-project.jetty=WARN' + os.linesep + \
                      'log4j.logger.org.spark-project.jetty.util.component.AbstractLifeCycle=ERROR' + os.linesep + \
                      'log4j.logger.org.apache.spark.repl.SparkIMain$exprTyper=INFO' + os.linesep + \
                      'log4j.logger.org.apache.spark.repl.SparkILoop$SparkILoopInterpreter=INFO' + os.linesep + \
                      'log4j.logger.com.vmeg.code=DEBUG'
    solved= log4jProperties.format(logFilePath=logFilePath, errorFilePath=errorFilePath)
    with open(outputFilePath,'w') as of:
        of.write(solved)
    return

def runJob(fileListPath,master,driverMemory,executorMemory,numExecutors,minSupport,numPartitions,cores,log4jFilePath,log4jExecPath,appname,isMinMin=0,isPFP=False,isFreq=False,isSong=False,setcover=0,collectStatistics=True):
    prefixcmd = 'spark-submit --class \"CanTreeMain\" ' \
          '--master {master} '\
          '--driver-memory {driverMemory} ' \
          '--executor-memory {executorMemory} ' \
          '--total-executor-cores {totalExecutorCores} ' \
          '--executor-cores {cores} ' \
          '--files {log4jExecPath} ' \
          '--conf \"spark.driver.extraJavaOptions=-Dlog4j.configuration=file:{log4jFilePath} -Xss1g\" ' \
          '--conf \"spark.executor.extraJavaOptions=-Dlog4j.configuration=file:{log4jExecPath} -Xss1g\" '.format(master=master,
                                                                                                        driverMemory=driverMemory,
                                                                                                        executorMemory=executorMemory,
                                                                                                        totalExecutorCores=numExecutors*cores,
                                                                                                        cores=cores,
                                                                                                        log4jFilePath=log4jFilePath,
                                                                                                        log4jExecPath=log4jExecPath)
    jarFile =  'target/scala-2.11/cantree_2.11-0.1.jar'
    postcmd =' --num-partitions {numPartitions} ' \
             '--min-support {minSupport} ' \
             '--in-file-list-path {fileListPath} --app-name {appname} --collect-statistics {collectStatistics}'.format(numPartitions=numPartitions,
                                                         minSupport=minSupport,
                                                         fileListPath=fileListPath,appname=appname,collectStatistics=collectStatistics)
    if isPFP:
        postcmd += ' --pfp 1'
    if isFreq:
        postcmd += ' --freq-sort 1'
    if isSong:
        postcmd += ' --song 1'
    if isMinMin>0:
        postcmd += ' --min-min-support '+str(isMinMin)
    if setcover>0:
        postcmd += ' --set-cover '+str(setcover)
    cmdList = shlex.split(prefixcmd)
    cmdList.append(jarFile)
    cmdList+=(shlex.split(postcmd))
    try:
        print(cmdList)
        subprocess.check_output(cmdList)
    except subprocess.CalledProcessError as e:
        print("-Error- \n"+str(e.cmd))
        print(e.output)


def runTests(outputDir,
             log4jBaseDir,
             master,
             testFilePath,
             testCaseName,
             isFreq=False,
             minMinSup=0,
             pfp=False,
             song=False,
             setcover=0,
             minSupport = [0.001],
             coresExecutorMemNums = [(40,4,'20g','40g')],
             partitions = [1000],
             collectStatistics = True
             ):
    # minSupport = [0.001]
    log4jPath = 'src/main/resources/log4j_file.properties'
    log4jExecPath = 'src/main/resources/log4j_file_executor.properties'
    testCaseFiles = [(testFilePath,testCaseName)]
    for supp in minSupport:
        for cem in coresExecutorMemNums:
            cores,execNums,mem,driverMem = cem
            for partition in partitions:
                for testCase in testCaseFiles:
                    testFile,testName = testCase
                    testname = '_'.join([testName,str(partition),str(cores),str(execNums),mem,str(supp).replace('.','_')])
                    if collectStatistics:
                        testname = "STATIST_" + testname
                    if isFreq:
                        testname = "FREQ_" + testname
                    if minMinSup>0:
                        testname = "MINMIN_" + testname
                    if pfp:
                        testname = "PFP_"+testname
                    if song:
                        testname = "SONG_"+testname
                    if setcover>0:
                        testname = "SETCOVER_"+testname
                    log4jErrorFileName = os.path.join(outputDir,testname+'_error.txt')
                    execLog4jErrorFileName = os.path.join(outputDir,testname+'_error_exec.txt')
                    log4jFileName = os.path.join(outputDir,testname+'.txt')
                    execLog4jFileName = os.path.join(outputDir,testname+'_exec.txt')
                    prepareLogFile(log4jFileName,log4jErrorFileName,os.path.join(log4jBaseDir,log4jPath))
                    prepareLogFile(execLog4jFileName,execLog4jErrorFileName,os.path.join(log4jBaseDir,log4jExecPath))
                    if pfp:
                        runJob(testFile,master,driverMem,mem,execNums,supp,partition,cores,log4jPath,log4jExecPath,testname,minMinSup,True,isFreq,False,0,collectStatistics)
                    elif song:
                        runJob(testFile,master,driverMem,mem,execNums,supp,partition,cores,log4jPath,log4jExecPath,testname,minMinSup,False,False,song,0,collectStatistics)
                    elif setcover>0:
                        runJob(testFile,master,driverMem,mem,execNums,supp,partition,cores,log4jPath,log4jExecPath,testname,minMinSup,False,isFreq,False,setcover,collectStatistics)
                    else:
                        runJob(testFile,master,driverMem,mem,execNums,supp,partition,cores,log4jPath,log4jExecPath,testname,minMinSup,False,isFreq,False,0,collectStatistics)

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Run FIS test cases.')
    parser.add_argument('--outputdir', dest='outputdir', help='logs output dir')
    parser.add_argument('--log4jbasedir', dest='log4jBaseDir',help='base dir of cantree dir')
    parser.add_argument('--master', dest='master', help='master host')
    parser.add_argument('--testpath', dest='testpath', help='master host')
    parser.add_argument('--testname', dest='testname', help='master host')
    parser.add_argument('--freq', dest='freq', help='master host')
    parser.add_argument('--minmin', dest='minmin', help='Use frequency order')
    parser.add_argument('--pfp', dest='pfp', help='run regular pfp')
    parser.add_argument('--set-cover', dest='setcover', help='run using set cover technique')
    parser.add_argument('--song', dest='song', help='run song test case')
    # outputDir,log4jBaseDir,master
    args = parser.parse_args()
    runTests(args.outputdir,args.log4jBaseDir,args.master,args.testpath,args.testname,args.freq,args.minmin,args.pfp,args.song,args.setcover)




